### PR TITLE
[UI Toolkit] Add Class for verifying credentials in the editor

### PR DESCRIPTION
**Pull request by JakeCataford**
This class allows verifying credentials in the editor. It takes a
context object that implement IShopCredentials and is able to make
requests and mutate the verification state on that context object.

It also includes a delegate function for drawing the UI for verifying
credentials, it's designed to be initialized in OnEnable and called out
to in OnInspectorGUI

### CODE REVIEW

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                }
+            });
+        }
+
+        /// <summary>
+		/// Perform an async request to see if the credentials are valid.
+		/// This request will also mutate the context object's CredentialsVerificationState.
+        /// </summary>
+        public void VerifyCredentials() {
+            VerifyCredentials(() => { }, () => { });
```
Comment (mikkoh) [inline Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: What's going on here?

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+		/// <summary>
+		/// Have the context's credentials been verified already?
+		/// </summary>
+		/// <returns>True if the context's credentials are verified</returns>
+		public bool HasVerifiedCredentials() {
+            return _context.CredentialsVerificationState == ShopCredentialsVerificationState.Verified;
```
Comment (mikkoh) [inline Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: Should this be:
```
return ! _isRequestInProgress && _context.CredentialsVerificationState == ShopCredentialsVerificationState.Verified;
```

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs (hunk 1 above)
Comment (JakeCataford) [reply Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: Just an overload with empty onSuccess/onError callbacks, those are useful for tests/the editor but internally its easier to just call this overload

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs (hunk 2 above)
Comment (JakeCataford) [reply Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: Im worried that combining these states might make a weird situation/a janky editor later on... when the verification request is ongoing, we can trust the last verified state to render UI. But if we want to change the ui when a request is in progress we need to have a getter that tells us a request is in progress... We can create that getter when we need it in an editor?

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+        /// <summary>
+        /// Perform an async request to see if the credentials are valid.
+        /// This request will also mutate the context object's CredentialsVerificationState.
+        /// </summary>
+        /// <param name="onSuccess">Callback called when the credentials are found to be valid</param>
+        /// <param name="onFailure">Callback called when the credentials are found to be invalid</param>
+        public void VerifyCredentials(Action onSuccess, Action onFailure) {
+            if (_isRequestInProgress) 
+                throw new InvalidOperationException("Can't verify credentials when a verification request is already in progress.");
```
Comment (sleroux) [inline Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: What does it mean to throw an exception in the context of an editor script? Does it just exit or lock up anything?

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+            _isRequestInProgress = true;
+
+            Client().Query((root) => {
+                root.shop((shop) => {
```
Comment (sleroux) [inline Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: Nit: There's a default query for this `DefaultQueries.shop.Name(query)`

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs (hunk 3 above)
Comment (JakeCataford) [reply Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: its similar to how it works during the game: the execution context of the script dies and it effectively stops OnInspectorGUI or the Update callback at that point so it wont kill the editor or anything nefarious

**Comment by JakeCataford** [comment]
@mikkoh wdyt of my response? they got clobbered when i reformatted the files

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs (hunk 2 above)
Comment (mikkoh) [reply Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: :+1:

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    /// </summary>
+    public class ShopCredentialsVerifier {
+        private IShopCredentials _context;
+        private bool _isRequestInProgress;
+
+        /// <summary>
+        /// Creates a new verifier
+        /// </summary>
+        /// <param name="context">The object with credentials</param>
+        public ShopCredentialsVerifier(IShopCredentials context) {
```
Comment (mikkoh) [inline Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: `context` is weird naming to me. Can we change to `credentials`

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    using Shopify.Unity;
+    using Shopify.Unity.SDK;
+    using Shopify.Unity.SDK.Editor;
+
+    /// <summary>
+    /// A helper for the editor that allows verifying shop credentials
+    /// and drawing a UI to support the feature.
+    /// </summary>
+    public class ShopCredentialsVerifier {
+        private IShopCredentials _context;
```
Comment (mikkoh) [inline Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: We've tried to follow the C# way of using Pascal Case for member variables. Can we change to `_Context`. Well actually maybe `_Credentials`.

File: Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs (hunk 5 above)
Comment (JakeCataford) [reply Assets/Shopify/UIToolkit/Editor/ShopCredentialsVerifier.cs]: sure thing.

**Comment by JakeCataford** [comment]
@mikkoh addressed these changes

**Pull request merged**